File: 1-realtime-location-updates-at-scale/0-typescript/src/location/location.controller.ts
import { BadRequestException, Body, Controller, Get, Post, Query } from "@nestjs/common"
import { LocationService } from "./location.service"

interface UpdateBody {
    driverId: string
    lat: number
    lng: number
}

@Controller()
export class LocationController {
    constructor(private readonly location: LocationService) {}

    @Get("/api/health")
    async health(): Promise<{ status: string }> {
        const ok = await this.location.ping()
        return { status: ok ? "ok" : "down" }
    }

    @Post("/api/location/update")
    async update(@Body() body: UpdateBody): Promise<{ driverId: string; updated: boolean }> {
        const { driverId, lat, lng } = body
        if (!driverId || !this.isLat(lat) || !this.isLng(lng)) {
            throw new BadRequestException("invalid driverId or coordinates")
        }
        await this.location.updateLocation(driverId, lat, lng)
        return { driverId, updated: true }
    }

    @Get("/api/location/nearby")
    async nearby(
        @Query("lat") lat: string,
        @Query("lng") lng: string,
        @Query("radius") radius: string,
    ): Promise<{ center: { lat: number; lng: number }; radiusM: number; drivers: unknown[] }> {
        const latN = Number(lat)
        const lngN = Number(lng)
        const radiusM = Number(radius)
        if (!this.isLat(latN) || !this.isLng(lngN) || !(radiusM > 0)) {
            throw new BadRequestException("lat, lng and positive radius are required")
        }
        const drivers = await this.location.findNearby(latN, lngN, radiusM)
        return { center: { lat: latN, lng: lngN }, radiusM, drivers }
    }

    private isLat(v: number): boolean {
        return Number.isFinite(v) && v >= -90 && v <= 90
    }

    private isLng(v: number): boolean {
        return Number.isFinite(v) && v >= -180 && v <= 180
    }
}

File: 1-realtime-location-updates-at-scale/1-java/src/main/java/com/starci/locationtracker/LocationController.java
package com.starci.locationtracker;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@RestController
public class LocationController {

    private final LocationService location;

    public LocationController(LocationService location) {
        this.location = location;
    }

    @GetMapping("/api/health")
    public Map<String, String> health() {
        return Map.of("status", location.ping() ? "ok" : "down");
    }

    @PostMapping("/api/location/update")
    public ResponseEntity<?> update(@RequestBody UpdateRequest body) {
        if (body.driverId() == null || body.driverId().isBlank()
                || body.lat() < -90 || body.lat() > 90
                || body.lng() < -180 || body.lng() > 180) {
            return ResponseEntity.badRequest().body(Map.of("error", "invalid driverId or coordinates"));
        }
        location.updateLocation(body.driverId(), body.lat(), body.lng());
        return ResponseEntity.ok(Map.of("driverId", body.driverId(), "updated", true));
    }

    @GetMapping("/api/location/nearby")
    public ResponseEntity<?> nearby(@RequestParam double lat,
                                    @RequestParam double lng,
                                    @RequestParam double radius) {
        if (radius <= 0 || lat < -90 || lat > 90 || lng < -180 || lng > 180) {
            return ResponseEntity.badRequest().body(Map.of("error", "lat, lng and positive radius are required"));
        }
        List<LocationService.NearbyDriver> drivers = location.findNearby(lat, lng, radius);
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("center", Map.of("lat", lat, "lng", lng));
        response.put("radiusM", radius);
        response.put("drivers", drivers);
        return ResponseEntity.ok(response);
    }

    public record UpdateRequest(String driverId, double lat, double lng) {
    }
}
